src/digest/halfsiphash.cr

```diff
@@ -0,0 +1,201 @@
+#
+# Example:
+# ```
+# key = uninitialized Digest::HalfSipHash::Key
+# Random::Secure.random_bytes(key.to_slice)
+#
+# hasher = Digest::HalfSipHash(2, 4).new(key)
+# hasher.update("some ")
+# hasher.update("input data")
+# hash = hasher.result
+# ```
+struct Digest::HalfSipHash(CROUNDS, DROUNDS)
+  alias Key = StaticArray(UInt8, 8)
+
+  # :nodoc:
+  BUF_SIZE = sizeof(UInt32)
+
+  def initialize(key : Key)
+    @buf = uninitialized UInt8[4] # UInt8[BUF_SIZE]
+    @buf_index = 0
+    @inlen = 0_u32
+
+    k0 = u8to32_le(key.to_unsafe)
+    k1 = u8to32_le(key.to_unsafe + 4)
+
+    @v0 = 0_u32
+    @v1 = 0_u32
+    @v2 = 0x6c796765_u32
+    @v3 = 0x74656462_u32
+
+    @v3 ^= k1
+    @v2 ^= k0
+    @v1 ^= k1
+    @v0 ^= k0
+
+    @v1 ^= 0xee
+  end
+
+  def update(data : String) : Nil
+    update(data.to_unsafe, data.bytesize)
+  end
+
+  def update(data : Bytes) : Nil
+    update(data.to_unsafe, data.size)
+  end
+
+  protected def update(data : UInt8*, datalen : Int32) : Nil
+    @inlen += datalen
+    done = data + datalen
+    v0, v1, v2, v3 = @v0, @v1, @v2, @v3
+
+    unless @buf_index == 0
+      # fill incomplete 4-byte buffer
+      count = Math.min(datalen, BUF_SIZE - @buf_index)
+      data.copy_to(@buf.to_unsafe + @buf_index, count)
+
+      size = @buf_index + count
+      unless size == BUF_SIZE
+        @buf_index = size
+        return
+      end
+
+      # compress 4-byte buffer
+      update(@buf.to_unsafe)
+
+      @buf_index = 0
+      data += count
+      datalen -= count
+    end
+
+    # compress has many 4-bytes as possible
+    stop = data + (datalen - datalen % BUF_SIZE)
+    until data == stop
+      update(data)
+      data += BUF_SIZE
+    end
+
+    # save incomplete 4-byte to buffer
+    unless stop == done
+      @buf_index = left = datalen & 3
+      data.copy_to(@buf.to_unsafe, left)
+    end
+
+    @v0, @v1, @v2, @v3 = v0, v1, v2, v3
+  end
+
+  private macro update(input)
+    m = u8to32_le({{input}})
+    v3 ^= m
+
+    CROUNDS.times { sipround }
+
+    v0 ^= m
+  end
+
+  def result
+    output = uninitialized UInt64
+    result(@buf.to_unsafe, @buf_index, pointerof(output).as(UInt8*))
+    output
+  end
+
+  def result(output : Bytes) : Nil
+    raise ArgumentError.new("Digest::HalfSipHash can only generate 8 bytes hashes.") unless output.size == 8
+    result(@buf.to_unsafe, @buf_index, output.to_unsafe)
+  end
+
+  protected def result(input : UInt8*, left : Int32, output : UInt8*) : Nil
+    b = @inlen << 24
+    v0, v1, v2, v3 = @v0, @v1, @v2, @v3
+
+    case left
+    when 3
+      b |= input[2].to_u32 << 16
+      b |= input[1].to_u32 << 8
+      b |= input[0].to_u32
+    when 2
+      b |= input[1].to_u32 << 8
+      b |= input[0].to_u32
+    when 1
+      b |= input[0].to_u32
+    end
+
+    v3 ^= b
+
+    CROUNDS.times { sipround }
+
+    v0 ^= b
+    v2 ^= 0xee
+
+    DROUNDS.times { sipround }
+
+    b = v1 ^ v3
+    u32to8_le(output, b)
+
+    v1 ^= 0xdd
+
+    DROUNDS.times { sipround }
```

Comment:
There's no need in whole second finalization. Just increase DROUNDS by 1, and use v1 and v3 separately, without xor.

Given hash-value is not going to be exposed, there is no need in being such strong.

Reply:
I follow the official implementation for halfsiphash to generate an UInt64. I'm not competent enough to change the algorithm.

Reply:
Please, believe me at least once.

You've aleady mentioned that `#hash` should not be exposed. With this limitation my suggestion is absolutely safe. Just believe me.

Reply:
Please don't do it, it's not safe. Just believe me.

@funny-falcon Actually, I do believe you. But if two people come with contradictory arguments and each say "Please believe me" there's no way to tell who is right, and why we should believe one or the other.

Can you provide a proof that doing that is safe?

Reply:
@asterite, its your language, do what you want.
I have no degree in cryptography, still I understand what I'm talking about.
You are just joking.
Nobody uses 32bit now, so this is irrelevant. Just add empty loop here, and no one will complain.

Reply:
I mean, no one will complain if this function will be much slower that it needs to be.

Reply:
I really have nothing against you! But I'm really incompetent to judge.

Those changes also mean that Digest::HalfSipHash would no longer be compliant to the specs, right? Since I want to expose it to developers, then the implementation should be compliant.

Since generating an UInt32 is okayish, maybe we can just return an UInt32 in Digest::HalfSipHash, and merely expand it to an UInt64 in Crystal::Hasher, so we don't have to deal with per-platform types, yet keep the algorithm compliant?

Reply:
I think yes. Hash is for example used in the compiler, and @RX14 benchmarked the compiler and with this change (I think) String#hash takes 10~15% of the time. So a fast hash function is important. But together with that we should probably optimize `Hash` entirely, avoid invoking `#hash` for small hashes (just do a linear lookup). And maybe even have a specialized Hash for the compiler where order of insertion doesn't matter.

There's just so much to do...

Reply:
>  and merely expand it to an UInt64 in Crystal::Hasher

I think, it is good option. Simple multiplication by non-trivial odd number does the trick.